src/rust-crypto/dehydration.ts

```diff
@@ -0,0 +1,262 @@
+        crypto.getRandomValues(key);
+        await this.secretStorage.store(SECRET_STORAGE_NAME.name, encodeUnpaddedBase64(key));
+        this.key = key;
+    }
+
+    /** Get and cache the encryption key from secret storage
+     *
+     * If `create` is `true`, creates a new key if no existing key is present.
+     */
+    private async getKey(create: boolean): Promise<boolean> {
+        if (this.key === undefined) {
+            const keyB64 = await this.secretStorage.get(SECRET_STORAGE_NAME.name);
+            if (keyB64 === undefined) {
+                if (!create) {
+                    return false;
+                }
+                await this.resetKey();
+            } else {
+                this.key = decodeBase64(keyB64);
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Rehydrate the dehydrated device stored on the server
+     *
+     * Checks if there is a dehydrated device on the server.  If so, rehydrates
+     * the device and processes the to-device events.
+     *
+     * Returns whether or not a dehydrated device was found.
+     */
+    public async rehydrateDeviceIfAvailable(): Promise<boolean> {
+        if (!(await this.getKey(false))) {
+            return false;
+        }
+
+        let dehydratedDeviceResp;
+        try {
+            dehydratedDeviceResp = await this.http.authedRequest<IDehydratedDeviceResp>(
+                Method.Get,
+                "/dehydrated_device",
+                undefined,
+                undefined,
+                {
+                    prefix: UnstablePrefix,
+                },
+            );
+        } catch (error) {
+            const err = error as MatrixError;
+            // We ignore M_NOT_FOUND (there is no dehydrated device, so nothing
+            // us to do) and M_UNRECOGNIZED (the server does not understand the
+            // endpoint).  We pass through any other errors.
+            if (err.errcode === "M_NOT_FOUND" || err.errcode === "M_UNRECOGNIZED") {
+                this.logger.info("dehydration: No dehydrated device");
+                return false;
+            }
+            throw err;
+        }
+
+        this.logger.info("dehydration: device found");
+
+        const rehydratedDevice = await this.olmMachine
+            .dehydratedDevices()
+            .rehydrate(
+                this.key!,
+                new RustSdkCryptoJs.DeviceId(dehydratedDeviceResp.device_id),
+                JSON.stringify(dehydratedDeviceResp.device_data),
+            );
+
+        this.logger.info("dehydration: device rehydrated");
+
+        let nextBatch: string | undefined = undefined;
+        let toDeviceCount = 0;
+        let roomKeyCount = 0;
+        for (;;) {
+            const path = encodeUri("/dehydrated_device/$device_id/events", {
+                $device_id: dehydratedDeviceResp.device_id,
+            });
+            const eventResp: IDehydratedDeviceEventsResp = await this.http.authedRequest<IDehydratedDeviceEventsResp>(
```

Comment:
I don't think the type annotation is needed here?

```suggestion
            const eventResp = await this.http.authedRequest<IDehydratedDeviceEventsResp>(
```

Reply:
If I omit the type annotation, the TypeScript checker says "error TS7022: 'eventResp' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer."

I don't understand why, because other calls to `authedRequest` don't need the type annotation.

Reply:
:shrug: